File: examples/kotlin/flappy_bird/src/main/kotlin/BirdAnimator.kt
import com.goudengine.core.GoudGame

class BirdAnimator(private val game: GoudGame) {
    private val frameTextures = mutableListOf<Long>()
    private var currentFrame = 0
    private var animationTime = 0f
    private val frameDuration = 0.1f

    private var currentX = 0f
    private var currentY = 0f
    private var currentRotation = 0f

    companion object {
        const val BIRD_WIDTH = 34f
        const val BIRD_HEIGHT = 24f
    }

    fun initialize() {
        frameTextures.add(game.loadTexture("assets/sprites/bluebird-downflap.png"))
        frameTextures.add(game.loadTexture("assets/sprites/bluebird-midflap.png"))
        frameTextures.add(game.loadTexture("assets/sprites/bluebird-upflap.png"))
    }

    fun update(deltaTime: Float, x: Float, y: Float, rotation: Float) {
        animationTime += deltaTime
        if (animationTime >= frameDuration) {
            currentFrame = (currentFrame + 1) % frameTextures.size
            animationTime = 0f
        }
        currentX = x
        currentY = y
        currentRotation = rotation
    }

    fun draw() {
        if (frameTextures.isEmpty()) return
        game.drawSprite(
            frameTextures[currentFrame],
            currentX + BIRD_WIDTH / 2f,
            currentY + BIRD_HEIGHT / 2f,
            BIRD_WIDTH,
            BIRD_HEIGHT,
            currentRotation * Math.PI.toFloat() / 180f
        )
    }

    fun reset() {
        currentFrame = 0
        animationTime = 0f
        currentRotation = 0f
    }
}

File: examples/ios/flappy_bird/Sources/FlappyBird/BirdAnimator.swift
// BirdAnimator.swift -- iOS Flappy Bird

import Foundation
import GoudEngine

final class BirdAnimator {
    private let game: GoudGame
    private var frameTextures: [UInt64] = []

    private var currentFrame: Int = 0
    private var animationTime: Float = 0
    private let frameDuration: Float

    private var initialX: Float
    private var initialY: Float
    private var currentX: Float
    private var currentY: Float
    private var currentRotation: Float = 0

    private let birdWidth: Float = 34
    private let birdHeight: Float = 24

    init(game: GoudGame, x: Float, y: Float, frameDuration: Float = 0.1) {
        self.game = game
        self.frameDuration = frameDuration
        self.initialX = x
        self.initialY = y
        self.currentX = x
        self.currentY = y
    }

    func initialize() {
        frameTextures = [
            game.loadTexture(path: "assets/sprites/bluebird-downflap.png"),
            game.loadTexture(path: "assets/sprites/bluebird-midflap.png"),
            game.loadTexture(path: "assets/sprites/bluebird-upflap.png"),
        ]
    }

    func update(deltaTime: Float, x: Float, y: Float, rotation: Float) {
        animationTime += deltaTime
        if animationTime >= frameDuration {
            currentFrame = (currentFrame + 1) % frameTextures.count
            animationTime = 0
        }
        currentX = x
        currentY = y
        currentRotation = rotation
    }

    func draw() {
        guard !frameTextures.isEmpty else { return }
        game.drawSprite(
            texture: frameTextures[currentFrame],
            x: currentX + birdWidth / 2,
            y: currentY + birdHeight / 2,
            width: birdWidth,
            height: birdHeight,
            rotation: currentRotation * Float.pi / 180
        )
    }

    func reset() {
        currentFrame = 0
        animationTime = 0
        currentX = initialX
        currentY = initialY
        currentRotation = 0
    }
}
